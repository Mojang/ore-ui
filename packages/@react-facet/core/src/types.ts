export type Immutable = boolean | number | string | undefined | null

export type ObjectWithImmutables = Record<string, Immutable>

export interface EqualityCheck<T> {
  (): (current: T) => boolean
}

/**
 * Possible values a facet can have.
 * Currently functions are not supported
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Value = string | number | boolean | undefined | null | [] | Record<string, any> | bigint

export type ExtractFacetValues<T extends ReadonlyArray<Facet<unknown>>> = {
  [K in keyof T]: T[K] extends Facet<infer V> ? V : never
}

export const FACET_FACTORY = Symbol('facet-factory')

export interface FacetFactory<T> {
  (): Facet<T>
  factory: typeof FACET_FACTORY
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ExcludeFacetFactory<T> = T extends FacetFactory<any> ? never : T

export interface Listener<T> {
  (value: T): void
}

export type Unsubscribe = () => void

export type Observe<T> = (listener: Listener<T>) => Unsubscribe

export interface Facet<T> {
  get: () => Option<T>
  observe: Observe<T>
}

export interface WritableFacet<T> extends Facet<T> {
  set: (value: T) => void
  setWithCallback: (callback: (previousValue: Option<T>) => Option<T>) => void
}

export interface Update<V> {
  (value: V): void
}

export type Setter<V> = (callbackSetter: V | ((previousValue: Option<V>) => Option<V>)) => void

export type Cleanup = Unsubscribe

export type StartSubscription<V> = (update: Update<V>) => Cleanup

export const isFacet = <T>(value: Value | Facet<T>): value is Facet<T> => {
  return (
    value !== null &&
    value !== undefined &&
    (value as Facet<T>).observe !== undefined &&
    (value as Facet<T>).get !== undefined
  )
}

export type FacetProp<T extends Value> = Facet<T> | ExcludeFacetFactory<T>

export interface Effect<T> {
  (value: T): void | Cleanup
}

export const NO_VALUE = Symbol.for('NoValue')

export type NoValue = typeof NO_VALUE

export type Option<T> = NoValue | T

/**
 * Unit of work that can be scheduled within a batch.
 */
export type Task = {
  (): void
  scheduled?: boolean
}

/**
 * Function that when executed will have its Facet updates scheduled.
 */
export type Batch = () => void

/**
 * CSSStyleDeclaration that has FacetDefinition as values and includes Gameface's extended number based properties
 * Based on Gameface's TypeScript definition: https://github.com/Mojang/Mojang_Cohtml/blob/master/src/Scripting/scripted/typescript/ScriptedCSSStyleDeclaration.d.ts
 */
export type FacetCSSStyleDeclaration = {
  alignContent?: FacetProp<string>
  alignItems?: FacetProp<string>
  alignSelf?: FacetProp<string>
  animation?: FacetProp<string>
  animationDelay?: FacetProp<string>
  animationDirection?: FacetProp<string>
  animationDuration?: FacetProp<string>
  animationFillMode?: FacetProp<string>
  animationIterationCount?: FacetProp<string>
  animationName?: FacetProp<string>
  animationPlayState?: FacetProp<string>
  animationTimingFunction?: FacetProp<string>
  background?: FacetProp<string>
  backgroundColor?: FacetProp<string>
  backgroundImage?: FacetProp<string>
  backgroundPosition?: FacetProp<string>
  backgroundRepeat?: FacetProp<string>
  backgroundSize?: FacetProp<string>
  border?: FacetProp<string>
  borderBottom?: FacetProp<string>
  borderBottomColor?: FacetProp<string>
  borderBottomLeftRadius?: FacetProp<string>
  borderBottomRightRadius?: FacetProp<string>
  borderBottomStyle?: FacetProp<string>
  borderBottomWidth?: FacetProp<string>
  borderColor?: FacetProp<string>
  borderImage?: FacetProp<string>
  borderImageOutset?: FacetProp<string>
  borderImageRepeat?: FacetProp<string>
  borderImageSlice?: FacetProp<string>
  borderImageSource?: FacetProp<string>
  borderImageWidth?: FacetProp<string>
  borderLeft?: FacetProp<string>
  borderLeftColor?: FacetProp<string>
  borderLeftStyle?: FacetProp<string>
  borderLeftWidth?: FacetProp<string>
  borderRadius?: FacetProp<string>
  borderRight?: FacetProp<string>
  borderRightColor?: FacetProp<string>
  borderRightStyle?: FacetProp<string>
  borderRightWidth?: FacetProp<string>
  borderStyle?: FacetProp<string>
  borderTop?: FacetProp<string>
  borderTopColor?: FacetProp<string>
  borderTopLeftRadius?: FacetProp<string>
  borderTopRightRadius?: FacetProp<string>
  borderTopStyle?: FacetProp<string>
  borderTopWidth?: FacetProp<string>
  borderWidth?: FacetProp<string>
  bottom?: FacetProp<string>
  bottomPX?: FacetProp<number>
  bottomVH?: FacetProp<number>
  bottomVW?: FacetProp<number>
  boxShadow?: FacetProp<string>
  clipPath?: FacetProp<string>
  color?: FacetProp<string>
  contain?: FacetProp<string>
  cursor?: FacetProp<string>
  display?: FacetProp<'flex' | 'none'>
  filter?: FacetProp<string>
  flex?: FacetProp<string>
  flexDirection?: FacetProp<string>
  flexWrap?: FacetProp<string>
  font?: FacetProp<string>
  fontFamily?: FacetProp<string>
  fontFitMode?: FacetProp<string>
  fontSize?: FacetProp<string>
  fontStyle?: FacetProp<string>
  fontWeight?: FacetProp<string>
  height?: FacetProp<string>
  heightPX?: FacetProp<number>
  heightVH?: FacetProp<number>
  isolation?: FacetProp<string>
  justifyContent?: FacetProp<string>
  left?: FacetProp<string>
  leftPX?: FacetProp<number>
  leftVH?: FacetProp<number>
  leftVW?: FacetProp<number>
  letterSpacing?: FacetProp<string>
  lineHeight?: FacetProp<string>
  margin?: FacetProp<string>
  marginBottom?: FacetProp<string>
  marginLeft?: FacetProp<string>
  marginRight?: FacetProp<string>
  marginTop?: FacetProp<string>
  mask?: FacetProp<string>
  maskClip?: FacetProp<string>
  maskImage?: FacetProp<string>
  maskMode?: FacetProp<string>
  maskPosition?: FacetProp<string>
  maskRepeat?: FacetProp<string>
  maskSize?: FacetProp<string>
  maxHeight?: FacetProp<string>
  maxHeightPX?: FacetProp<number>
  maxHeightVH?: FacetProp<number>
  maxWidth?: FacetProp<string>
  maxWidthPX?: FacetProp<number>
  maxWidthVW?: FacetProp<number>
  minHeight?: FacetProp<string>
  minHeightPX?: FacetProp<number>
  minHeightVH?: FacetProp<number>
  minWidth?: FacetProp<string>
  minWidthPX?: FacetProp<number>
  minWidthVW?: FacetProp<number>
  mixBlendMode?: FacetProp<string>
  opacity?: FacetProp<number>
  overflow?: FacetProp<string>
  overflowWrap?: FacetProp<string>
  overflowX?: FacetProp<string>
  overflowY?: FacetProp<string>
  padding?: FacetProp<string>
  paddingBottom?: FacetProp<string>
  paddingLeft?: FacetProp<string>
  paddingRight?: FacetProp<string>
  paddingTop?: FacetProp<string>
  perspective?: FacetProp<string>
  perspectiveOrigin?: FacetProp<string>
  pointerEvents?: FacetProp<string>
  position?: FacetProp<string>
  right?: FacetProp<string>
  rightPX?: FacetProp<number>
  rightVH?: FacetProp<number>
  rightVW?: FacetProp<number>
  textAlign?: FacetProp<string>
  textOverflow?: FacetProp<string>
  textShadow?: FacetProp<string>
  textTransform?: FacetProp<string>
  top?: FacetProp<string>
  topPX?: FacetProp<number>
  topVH?: FacetProp<number>
  topVW?: FacetProp<number>
  transform?: FacetProp<string>
  transformOrigin?: FacetProp<string>
  transition?: FacetProp<string>
  transitionDelay?: FacetProp<string>
  transitionDuration?: FacetProp<string>
  transitionProperty?: FacetProp<string>
  transitionTimingFunction?: FacetProp<string>
  userSelect?: FacetProp<string>
  visibility?: FacetProp<string>
  whiteSpace?: FacetProp<string>
  width?: FacetProp<string>
  widthPX?: FacetProp<number>
  widthVW?: FacetProp<number>
  zIndex?: FacetProp<number>
}
