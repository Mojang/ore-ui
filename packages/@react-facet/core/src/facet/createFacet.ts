import { defaultEqualityCheck } from '../equalityChecks'
import { Cleanup, EqualityCheck, Listener, WritableFacet, StartSubscription, Option, NO_VALUE } from '../types'
import { batch } from '../scheduler'

export interface FacetOptions<V> {
  initialValue: Option<V>
  startSubscription?: StartSubscription<V>
  equalityCheck?: EqualityCheck<V>
}
/**
 * The low level function to create a Facet, not recommended to be used if you can use any of the react facet hooks to create facets instead (Ex: useFacetState, useFacetWrap)
 */
export function createFacet<V>({
  initialValue,
  startSubscription,
  equalityCheck = defaultEqualityCheck,
}: FacetOptions<V>): WritableFacet<V> {
  const listeners: Set<Listener<V>> = new Set()
  let currentValue = initialValue
  let cleanupSubscription: Cleanup | undefined

  const checker = equalityCheck?.()

  const update = (newValue: V) => {
    if (equalityCheck !== undefined) {
      // we optimize for the most common scenario of using the defaultEqualityCheck (by inline its implementation)
      if (equalityCheck === defaultEqualityCheck) {
        const typeofValue = typeof newValue
        if (
          (typeofValue === 'number' ||
            typeofValue === 'string' ||
            typeofValue === 'boolean' ||
            newValue === null ||
            newValue === undefined) &&
          currentValue === newValue
        ) {
          return
        }
      } else {
        if (checker !== undefined && checker(newValue)) {
          return
        }
      }
    }

    batch(() => {
      currentValue = newValue

      for (const listener of listeners) {
        listener(currentValue)
      }
    })
  }

  /**
   * Simpler update implementation that only resets the value and runs all cleanup functions.
   * Done as a separated function to not interfere with the usual "hot-path" of the update function.
   */
  const updateToNoValue = () => {
    currentValue = NO_VALUE
  }

  return {
    set: update,

    setWithCallback: (setter) => {
      const value = setter(currentValue)
      if (value === NO_VALUE) {
        updateToNoValue()
      } else {
        update(value)
      }
    },

    get: () => currentValue,

    observe: (listener) => {
      listeners.add(listener)

      if (currentValue !== NO_VALUE) {
        listener(currentValue)
      }

      // This is the first subscription, so we start subscribing to dependencies
      if (listeners.size === 1 && startSubscription) {
        cleanupSubscription = startSubscription(update)
      }

      return () => {
        listeners.delete(listener)

        // if this was the last to unsubscribe, we unsubscribe from our dependencies
        if (listeners.size === 0 && cleanupSubscription) {
          currentValue = initialValue
          cleanupSubscription()
        }
      }
    },
  }
}
