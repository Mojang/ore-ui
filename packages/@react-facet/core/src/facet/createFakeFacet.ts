import { Facet } from '..'
/**
 * Creates a nonwritable barebones facet to be used when you need an initial facet value that's meant to be replaced later by a real facet. Ex: with `useContext()`
 */
export function createFakeFacet<T>(value: T): Facet<T> {
  const facet: Facet<T> = {
    get: () => value,
    observe: (listener) => {
      if (process.env.NODE_ENV !== 'production') {
        console.log(`Accessing a fake facet, perhaps you're missing a Context Provider?`)
      }
      listener(value)
      return () => {}
    },
  }
  return facet
}
