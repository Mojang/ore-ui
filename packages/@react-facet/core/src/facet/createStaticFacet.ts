import { Facet } from '../types'
/**
 * Creates a nonwritable barebones static facet to be used when you need an initial facet value outside the react context
 * that's meant to be replaced later by a real facet. Ex: with `createContext()`
 */
export function createStaticFacet<T>(value: T): Facet<T> {
  const facet: Facet<T> = {
    get: () => value,
    observe: (listener) => {
      listener(value)
      return () => {}
    },
  }
  return facet
}
