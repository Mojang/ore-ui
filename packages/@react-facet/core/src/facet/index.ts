export * from './createFacet'
export * from './createFacet'
