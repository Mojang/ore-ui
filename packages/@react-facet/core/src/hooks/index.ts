export * from './useFacetCallback'
export * from './useFacetEffect'
export * from './useFacetLayoutEffect'
export * from './useFacetMap'
export * from './useFacetMemo'
export * from './useFacetPropSetter'
export * from './useFacetRef'
export * from './useFacetRef'
export * from './useFacetState'
export * from './useFacetUnwrap'
export * from './useFacetWrap'
