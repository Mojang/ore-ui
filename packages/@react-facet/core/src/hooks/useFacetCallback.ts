import { useCallback, useLayoutEffect, useRef } from 'react'
import { NoValue } from '..'
import { Facet, NO_VALUE, Option } from '../types'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V) => C,
  dependencies: unknown[],
  facet: [Facet<V>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, V4, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, V4, V5, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, V4, V5, V6, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>, Facet<V6>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, V4, V5, V6, V7, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>, Facet<V6>, Facet<V7>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, V4, V5, V6, V7, V8, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7, v8: V8) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>, Facet<V6>, Facet<V7>, Facet<V8>],
): C

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useFacetCallback<M, V, V1, V2, V3, V4, V5, V6, V7, V8, V9, C extends (...args: any[]) => M | NoValue>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7, v8: V8, v9: V9) => C,
  dependencies: unknown[],
  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>, Facet<V6>, Facet<V7>, Facet<V8>, Facet<V9>],
): C

export function useFacetCallback<
  M,
  V,
  V1,
  V2,
  V3,
  V4,
  V5,
  V6,
  V7,
  V8,
  V9,
  V10,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  C extends (...args: any[]) => M | NoValue,
>(
  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7, v8: V8, v9: V9, v10: V10) => C,
  dependencies: unknown[],
  facet: [
    Facet<V>,
    Facet<V1>,
    Facet<V2>,
    Facet<V3>,
    Facet<V4>,
    Facet<V5>,
    Facet<V6>,
    Facet<V7>,
    Facet<V8>,
    Facet<V9>,
    Facet<V10>,
  ],
): C

/**
 * Creates a callback that depends on the value of a facet.
 * Very similar to `useCallback` from `React`
 *
 * @param callback function callback that receives the current facet values and the arguments passed to the callback
 * @param dependencies variable used by the callback that are available in scope (similar as dependencies of useCallback)
 * @param facets facets that the callback listens to
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 */
export function useFacetCallback<M>(
  callback: (...args: unknown[]) => (...args: unknown[]) => M | NoValue,
  dependencies: unknown[],
  facets: Facet<unknown>[],
): (...args: unknown[]) => M | NoValue {
  const facetsRef = useRef<Option<unknown>[]>(facets.map(() => NO_VALUE))

  useLayoutEffect(() => {
    const unsubscribes = facets.map((facet, index) => {
      return facet.observe((value) => {
        facetsRef.current[index] = value
      })
    })

    return () => {
      unsubscribes.forEach((unsubscribe) => unsubscribe())
    }
    // We care about each individual facet and if any is a different reference
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, facets)

  // We care about each individual dependency and if any is a different reference
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const callbackMemoized = useCallback(callback, dependencies)

  // eslint-disable-next-line react-hooks/exhaustive-deps
  return useCallback(
    (...args: unknown[]) => {
      const values = facetsRef.current

      for (const value of values) {
        if (value === NO_VALUE) return NO_VALUE
      }

      return callbackMemoized(...values)(...args)
    },
    [callbackMemoized, facetsRef],
  )
}
