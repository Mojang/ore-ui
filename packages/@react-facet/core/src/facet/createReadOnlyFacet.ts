import { Facet } from '../types'
import { createFacet, FacetOptions } from './createFacet'

export function createReadOnlyFacet<V>(options: FacetOptions<V>): Facet<V> {
  return createFacet(options)
}
