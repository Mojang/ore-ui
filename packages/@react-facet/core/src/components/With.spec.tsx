import React from 'react'
import { act, render } from '@react-facet/dom-fiber-testing-library'
import { With } from '.'
import { createFacet } from '../facet'
import { Facet, NO_VALUE } from '../types'
import { useFacetEffect, useFacetMap } from '../hooks'

it('renders when not null, passing down the information', () => {
  const userFacet = createFacet({ initialValue: { user: 'Zelda' } })

  const Content = ({ data }: { data: Facet<{ user: string }> }) => {
    return (
      <div>
        Hello <fast-text text={useFacetMap(({ user }) => user, [], [data])} />
      </div>
    )
  }

  const Example = () => {
    return <With data={userFacet}>{(dataFacet) => <Content data={dataFacet} />}</With>
  }

  const scenario = <Example />

  const { container } = render(scenario)

  expect(container.firstChild).toMatchSnapshot()
})

it('does not render when nullish', () => {
  const userFacet = createFacet({ initialValue: null })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Hello there</div>
  }

  const Example = () => {
    return <With data={userFacet}>{() => <Content />}</With>
  }

  const scenario = <Example />

  render(scenario)

  expect(rendered).not.toHaveBeenCalled()
})

it('does not render facet has no value', () => {
  const userFacet = createFacet<boolean>({ initialValue: NO_VALUE })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Hello there</div>
  }

  const Example = () => {
    return <With data={userFacet}>{() => <Content />}</With>
  }
  const scenario = <Example />

  render(scenario)

  expect(rendered).not.toHaveBeenCalled()
})

it('correctly handles unmounting', () => {
  const mockFacet = createFacet<string | null>({ initialValue: 'abc' })

  const Content = ({ data }: { data: Facet<string> }) => {
    useFacetEffect(
      (data) => {
        if (data === null || data === undefined) {
          throw new Error('data should not be null')
        }
      },
      [],
      [data],
    )

    return <>mounted</>
  }

  const Example = () => {
    return <With data={mockFacet}>{(mock) => <Content data={mock} />}</With>
  }
  const scenario = <Example />

  const result = render(scenario)

  expect(result.container).toHaveTextContent('mounted')

  act(() => mockFacet.set(null))

  expect(result.container).not.toHaveTextContent('mounted')
})
