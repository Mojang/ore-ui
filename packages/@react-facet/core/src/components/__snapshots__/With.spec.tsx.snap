// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when not null, passing down the information 1`] = `
<div>
  Hello 
  Zelda
</div>
`;
